packages/network/src/index.ts

```diff
@@ -1,3 +1,19 @@
-type Fetch = typeof fetch;
+import type { BuildFetchOptions, Fetch, Middleware } from "./types";
 
-export type Middleware = (request: Request, fetch: Fetch) => ReturnType<Fetch>;
+import { Request, fetch } from 'cross-fetch';
+
+export function buildFetch(middleware: Middleware[], options?: BuildFetchOptions): Fetch {
+    return (input: RequestInfo, init?: RequestInit) => {
+        const request = new Request(input, init);
+
+        async function chain(request: Request, middleware: Middleware[]): Promise<Response> {
+            if (middleware.length === 0) return fetch(request)
+
+            return middleware[0](request, (input: RequestInfo, init?: RequestInit) => chain(request, middleware.slice(1)))
+        }
+
+        return chain(request, middleware)
+    }
```

Comment:
Any reason to recurse instead of looping?  This is harder to read (at least for me)

Reply:
updated as loop.